xdsl/backend/csl/lowering/convert_stencil_to_csl.py

```diff
@@ -0,0 +1,399 @@
+    inner_loop_body.add_ops(
+        [
+            # inner loop is y_id
+            y_id := SignednessCastOp.get_u(inner_loop_body.args[0]),
+            # compute boolean expression is_border_region_pe
+            pattern_minus_one := MinUI(pattern_op, one_u),
+            width_minus_xid := MinUI(width, x_id),
+            height_minus_yid := MinUI(height, y_id),
+            first := Cmpi(x_id, pattern_minus_one, "ult"),
+            second := Cmpi(y_id, pattern_minus_one, "ult"),
+            third := Cmpi(width_minus_xid, pattern_op, "ult"),
+            fourth := Cmpi(height_minus_yid, pattern_op, "ult"),
+            or_one := OrI(first, second),
+            or_two := OrI(or_one, third),
+            is_border_pe := OrI(or_two, fourth),
+            # generates: memcpy.get_params(xId)
+            memcpy_params := MemberCallOp.get(
+                "get_params", ComptimeStructType(), memcpy, x_id
+            ),
+            # generates: routes.computeAllRoutes(x_id, y_id, width, height, pattern)
+            route_params := MemberCallOp.get(
+                "computeAllRoutes",
+                ComptimeStructType(),
+                routes,
+                x_id,
+                y_id,
+                width,
+                height,
+                pattern_op,
+            ),
+            # setting up param structs for `@set_tile_code`
+            set_tile_params := ConstStructOp.get(
+                (Named.memcpy_params, memcpy_params),
+                (Named.stencil_comms_params, route_params),
+            ),
+            params_task := ConstStructOp.get((Named.is_border_region_pe, is_border_pe)),
+            set_tile_params_ext := ConcatStructOp.get(params_task, set_tile_params),
+            # generates: @set_tile_code(xId, yId, "pe.csl", .. param structs ..)
+            SetTileCodeOp.get(
+                ctx.program_module.sym_name,
+                x_id,
+                y_id,
+                set_tile_params_ext,
+            ),
+            Yield(),
+        ]
+    )
+    outer_loop_body.add_op(Yield())
+
+
+@dataclass(frozen=True)
+class ConvertStencilToCsl(ModulePass):
+    name = "convert-stencil-to-csl"
+
+    def apply(self, ctx: MLContext, op: ModuleOp) -> None:
+        # initialise both CSL modules
+        t_ctx = TranslationContext()
+        t_ctx.program_module = CslModuleOp(
+            regions=[Region(Block())],
+            properties={"kind": ModuleKindAttr(ModuleKind.PROGRAM)},
+            attributes={"sym_name": StringAttr("pe.csl")},
+        )
+        t_ctx.layout_module = CslModuleOp(
+            regions=[Region(Block())],
+            properties={"kind": ModuleKindAttr(ModuleKind.LAYOUT)},
+            attributes={"sym_name": StringAttr("layout.csl")},
+        )
+        # iterate over stencil.apply ops to find the maximum stencil width ('arm length')
+        assert len(op.body.block.ops) == 1
+        assert isinstance(stencil_func := op.body.block.first_op, func.FuncOp)
```

Comment:
I don't think asserting something this strong about program structure is a great idea, as it will just crash in most cases.